core/src/main/scala/akka/kafka/KafkaConsumerActor.scala

```diff
@@ -200,42 +227,46 @@ private[kafka] class KafkaConsumerActor[K, V](settings: ConsumerSettings[K, V])
         while (i > 0 && commitsInProgress > 0) {
           LockSupport.parkNanos(10 * 1000)
           val pollTimeout = if (i == 1) 1L else 0L
-          checkNoResult(consumer.poll(pollTimeout))
+          checkNoResult(tryPoll(pollTimeout))
           i -= 1
         }
       }
       finally wakeupTask.cancel()
 
     }
     else {
-      val rawResult = try consumer.poll(pollTimeout().toMillis) finally wakeupTask.cancel()
-      if (!rawResult.isEmpty) {
-        //check the we got only requested partitions and did not drop any messages
-        val fetchedTps = rawResult.partitions().asScala
-        if ((fetchedTps diff partitionsToFetch).nonEmpty)
-          throw new IllegalArgumentException(s"Unexpected records polled. Expected: $partitionsToFetch, " +
-            s"result: ${rawResult.partitions()}, consumer assignment: ${consumer.assignment()}")
-
-        //send messages to actors
-        requests.foreach {
-          case (ref, req) =>
-            //gather all messages for ref
-            val messages = req.topics.foldLeft[Iterator[ConsumerRecord[K, V]]](Iterator.empty) {
-              case (acc, tp) =>
-                val tpMessages = rawResult.records(tp).asScala.iterator
-                if (acc.isEmpty) tpMessages
-                else acc ++ tpMessages
-            }
-            if (messages.nonEmpty) {
-              ref ! Messages(req.requestId, messages)
-              requests -= ref
-            }
-        }
-      }
+      val result = try tryPoll(pollTimeout().toMillis) finally wakeupTask.cancel()
```

Comment:
+1, that's why the first version was `Try`-based. I'll try to go with a `NoPollResult` exception and see if it's better.